lib/ads/ad_manager.js

```diff
@@ -106,14 +130,112 @@ shaka.ads.AdManager = class extends shaka.util.FakeEventTarget {
 
 
   /**
-   * @param {string} assetKey
-   * @param {string} assetId
-   * @param {boolean} isLive
-   * @param {string=} backupUrl
+   * @override
+   * @export
+   */
+  initServerSide(adContainer, video, player) {
+    // Check that Client Side IMA SDK has been included
+    // NOTE: (window['google'] && google.ima) check for any
+    // IMA SDK, including SDK for Server Side ads.
+    // The 3rd check insures we have the right SDK:
+    // {google.ima.dai} is an object that's part of DAI IMA SDK
+    // but not SS SDK.
+    if (!window['google'] || !google.ima || !google.ima.dai) {
+      throw new shaka.util.Error(
+          shaka.util.Error.Severity.CRITICAL,
+          shaka.util.Error.Category.ADS,
+          shaka.util.Error.Code.SS_IMA_SDK_MISSING);
+    }
+
+    this.ssAdManager_ = new shaka.ads.ServerSideAdManager(
+        adContainer, video, player,
+        (e) => {
+          const event = /** @type {!shaka.util.FakeEvent} */ (e);
+          if (event && event.type) {
+            switch (event.type) {
+              case shaka.ads.AdManager.AD_STARTED:
+                this.stats_.incrementStarted();
+                break;
+              case shaka.ads.AdManager.AD_COMPLETE:
+                this.stats_.incrementPlayedCompletely();
+                break;
+              case shaka.ads.AdManager.AD_SKIPPED:
+                this.stats_.incrementSkipped();
+                break;
+            }
+          }
+          this.dispatchEvent(event);
+        });
+  }
+
+
+  /**
+   * @override
    * @export
    */
-  loadServerSideStream(assetKey, assetId, isLive, backupUrl) {
-    // TODO
+  loadServerSideVodStream(
+      adTagParameters, apiKey, authToken, contentSourceId,
```

Comment:
The main idea is to make the application not know about DAI. It also takes into account whether the SDK is available or not, so that the application does not have to double check. Do you think it is better to change it and that the application has to know about the DAI APIs?

Reply:
Yes, requiring an IMA object is fine here.

They already have to know about IMA to provide asset Id and other information, all of which is tightly coupled to IMA and could not being reused for a different ad SDK/signalling system.
The only way any of this will work is if the app has the IMA SDK included and knowingly relies on it, so we're not loosing anything by requiring an IMA object.
This is also consistent with what we do for CS ads.

We might have to revisit this design if we expand to other ad SDKs or signalling systems, although I doubt that we'll be able to create a truly agnostic system given how much specific info IMA DAI requires.

Reply:
OK, I changed it to require a IMA object. Thanks for your explanation